README.md
some test of NTL (Number Theory Library)


ChangeLog
2017-08-07  Mutsuo Saito

	* add files fo autotools


src/test_mat.cpp
#include <NTL/mat_GF2.h>
#include <NTL/matrix.h>
#include <NTL/GF2.h>
#include <NTL/version.h>

using namespace NTL;
using namespace std;

int main(int argc, char * argv[])
{
    Mat<GF2> mat;
    mat.SetDims(2, 2);
    for (int i = 0; i < 2; i++) {
        for (int j = 0; j < 2; j++) {
            cout << "mat.put(i, j, 1)" << endl;
            mat.put(i, j, 1);
            cout << "mat.get(i, j) = " << mat.get(i, j) << endl;
            cout << "mat.put(i, j, 0)" << endl;
            mat.put(i, j, 0);
            cout << "mat.get(i, j) = " << mat.get(i, j) << endl;
        }
    }
    mat.put(0,0,1);
    mat.put(0,1,1);
    mat.put(1,0,1);
    mat.put(1,1,1);
    cout << mat << endl;
}


src/order_test.cpp
#include <inttypes.h>
#include <iostream>
#include <iomanip>
#include <cstdlib>
#include <cerrno>
#include <NTL/mat_GF2.h>
#include <NTL/GF2.h>
#include <NTL/matrix.h>
#include <vector>
#include "calc_tvalue.hpp"

using namespace std;
using namespace NTL;

//#define DEBUG

namespace {
    Mat<GF2> Ident;
    bool make_mat(Mat<GF2>& mat, uint64_t num);
    size_t calc_order(Mat<GF2>& mat, size_t max);
}

int main(int argc, char * argv[])
{

    if (argc < 4) {
        cout << argv[0] << " s m max_tvalue" << endl;
        return -1;
    }
    errno = 0;
    int s = strtoul(argv[1], NULL, 10);
    int m = strtoul(argv[2], NULL, 10);
    int max_tvalue = strtoul(argv[3], NULL, 10);
    if (errno) {
        cout << " s, m and max_tvalue should be numbers" << endl;
        return -1;
    }
    cout << "# s = " << dec << s << endl;
    cout << "# m = " << dec << m << endl;
    cout << "# max_tvalue = " << dec << max_tvalue << endl;
    ident(Ident, m);
#if defined(DEBUG)
    cout << "main step 0.1" << endl;
#endif
    int64_t max = 1 << (m * m);
    Mat<GF2> mat;
    mat.SetDims(m, m);
    Mat<GF2> array[s];
#if defined(DEBUG)
    cout << "main step 0.2" << endl;
#endif
    vector<vector<int> >to;
    to.resize(max_tvalue + 1);
#if defined(DEBUG)
    cout << "main step 1" << endl;
#endif
    for (int i = 0; i <= max_tvalue; i++) {
        to[i].resize(max);
        for (int64_t j = 0; j < max; j++) {
            to[i][j] = 0;
        }
    }
#if defined(DEBUG)
    cout << "main step 2" << endl;
#endif
    for (int64_t cnt = 1; cnt < max; cnt++) {
        if (!make_mat(mat, cnt)) {
            continue;
        }
        int order = calc_order(mat, max - 1);
        if (order < s) {
            continue;
        }
        array[0] = mat;
        for (int j = 1; j < s; j++) {
            array[j] = array[j - 1] * mat;
        }
        int tvalue = calc_tvalue(array, s, max_tvalue);
        if (tvalue > max_tvalue || tvalue < 0) {
            continue;
        }
        to[tvalue][order] += 1;
        if (tvalue == 0 && order == max -1) {
            cout << "found!" << endl;
            cout << "cnt = " << dec << cnt << endl;
            cout << "tvalue = " << dec << tvalue << endl;
            cout << "order = " << dec << order << endl;
            cout << mat;
        }
    }
    cout << "(tvalue, order) = cnt" << endl;
    for (int i = 0; i <= max_tvalue; i++) {
        for (int64_t j = 0; j < max; j++) {
            if (to[i][j] == 0) {
                continue;
            }
            cout << "(" << dec << i << ","
                 << j << ") = " << to[i][j] << endl;
        }
    }
    return 0;
}

namespace {
    bool make_mat(Mat<GF2>& mat, uint64_t num)
    {
#if defined(DEBUG)
        cout << "start make_mat" << endl;
#endif
        //cout << "num = " << dec << num << endl;
        int m = mat.NumRows();
        const uint64_t mask = ~UINT64_C(0) >> (64 - m);
        for (int i = 0; i < m; i++) {
            uint64_t row = num & mask;
            if (row == 0) {
#if defined(DEBUG)
                cout << "end make_mat" << endl;
#endif
                return false;
            }
            for (int j = 0; j < m; j++) {
                //cout << "mat.put(" << i << "," << j << "," << (row & 1) << ")"
                //     << endl;
                mat.put(i, j, row & 1);
                row = row >> 1;
            }
            num = num >> m;
        }
        //cout << "make_mat 1:" << mat << endl;
        Mat<GF2> w = mat;
        if (gauss(w) != m) {
#if defined(DEBUG)
            cout << "end make_mat" << endl;
#endif
            return false;
        }
        //cout << "make_mat 2:\n" << mat << endl;
#if defined(DEBUG)
        cout << "end make_mat" << endl;
#endif
        return true;
    }

    size_t calc_order(Mat<GF2>& mat, size_t max)
    {
        Mat<GF2> w = mat;
        for (size_t i = 1; i <= max; i++) {
            w = w * mat;
            if (w == mat) {
                return i;
            }
        }
        return max;
    }
}


src/config.h
/* src/config.h.  Generated from config.h.in by configure.  */
/* src/config.h.in.  Generated from configure.ac by autoheader.  */

/* define if the compiler supports basic C++11 syntax */
#define HAVE_CXX11 1

/* define if the compiler supports basic C++14 syntax */
#define HAVE_CXX14 1

/* Define to 1 if you have the <inttypes.h> header file. */
#define HAVE_INTTYPES_H 1

/* Define to 1 if you have the `gf2x' library (-lgf2x). */
/* #undef HAVE_LIBGF2X */

/* Define to 1 if you have the `gmp' library (-lgmp). */
#define HAVE_LIBGMP 1

/* Define to 1 if you have the `ntl' library (-lntl). */
#define HAVE_LIBNTL 1

/* Define to 1 if you have the <memory.h> header file. */
#define HAVE_MEMORY_H 1

/* Define if g++ supports C++0x features. */
#define HAVE_STDCXX_0X /**/

/* Define to 1 if you have the <stdint.h> header file. */
#define HAVE_STDINT_H 1

/* Define to 1 if you have the <stdlib.h> header file. */
#define HAVE_STDLIB_H 1

/* Define to 1 if you have the <strings.h> header file. */
#define HAVE_STRINGS_H 1

/* Define to 1 if you have the <string.h> header file. */
#define HAVE_STRING_H 1

/* Define to 1 if you have the <sys/stat.h> header file. */
#define HAVE_SYS_STAT_H 1

/* Define to 1 if you have the <sys/types.h> header file. */
#define HAVE_SYS_TYPES_H 1

/* Define to 1 if you have the <unistd.h> header file. */
#define HAVE_UNISTD_H 1

/* Name of package */
#define PACKAGE "ntl-test1"

/* Define to the address where bug reports for this package should be sent. */
#define PACKAGE_BUGREPORT "saito@manieth.com"

/* Define to the full name of this package. */
#define PACKAGE_NAME "ntl-test1"

/* Define to the full name and version of this package. */
#define PACKAGE_STRING "ntl-test1 0.1"

/* Define to the one symbol short name of this package. */
#define PACKAGE_TARNAME "ntl-test1"

/* Define to the home page for this package. */
#define PACKAGE_URL ""

/* Define to the version of this package. */
#define PACKAGE_VERSION "0.1"

/* Define to 1 if you have the ANSI C header files. */
#define STDC_HEADERS 1

/* Version number of package */
#define VERSION "0.1"


src/calc_tvalue.hpp
#pragma once
#ifndef CALC_TVALUE_HPP
#define CALC_TVALUE_HPP

#include <NTL/matrix.h>
#include <NTL/GF2.h>
#include "combination.hpp"

int calc_tvalue(NTL::Mat<NTL::GF2> array[], int dim, int max_tvalue)
{
    using namespace std;
    using namespace NTL;

    uint32_t s = dim;
    uint32_t m = array[0].NumRows();
    Combination index(s);
    int t = 0;
    for (; t <= max_tvalue; t++) {
        index.reset(m - t);
#if defined(DEBUG)
        cout << "t = " << dec << t << endl;
#endif
        //vector<uint32_t> v;
        bool pass = true;
        for (;;) {
            Mat<GF2> v;
            v.SetDims(m - t, m);
#if defined(DEBUG)
            index.print(cout);
#endif
            int vrow = 0;
            for (uint32_t i = 0; i < s; ++i) {
                for (int j = 0; j < index[i]; ++j) {
                    v[vrow++] = array[i][j];
                }
            }
            if (gauss(v) != m-t) {
                pass = false;
                break;
            }
            bool hasNext = index.next();
            if (!hasNext) {
                break;
            }
        }
        if (pass) {
            return t;
        }
    }
    return -1;
}

#endif // CALC_TVALUE_HPP


src/order_test_fp.cpp
#include <inttypes.h>
#include <iostream>
#include <iomanip>
#include <cstdlib>
#include <cerrno>
#include <NTL/lzz_p.h>
//#include <NTL/zz_p.h>
//#include <NTL/matrix.h>
#include <vector>
#include "calc_tvalue_fp.hpp"

using namespace std;
using namespace NTL;

//#define DEBUG

namespace {
    bool make_mat(Mat<zz_p>& mat, uint64_t num);
    size_t calc_order(Mat<zz_p>& mat, size_t max);
}

int main(int argc, char * argv[])
{

    if (argc < 5) {
        cout << argv[0] << " p s m max_tvalue" << endl;
        return -1;
    }
    errno = 0;
    int p = strtoul(argv[1], NULL, 10);
    int s = strtoul(argv[2], NULL, 10);
    int m = strtoul(argv[3], NULL, 10);
    int max_tvalue = strtoul(argv[4], NULL, 10);
    if (errno) {
        cout << "p, s, m and max_tvalue should be numbers" << endl;
        return -1;
    }
    cout << "# p = " << dec << p << endl;
    cout << "# s = " << dec << s << endl;
    cout << "# m = " << dec << m << endl;
    cout << "# max_tvalue = " << dec << max_tvalue << endl;
    zz_p::init(p);
#if defined(DEBUG)
    cout << "main step 0.1" << endl;
#endif
    int64_t max = pow(p, m * m);
    Mat<zz_p> mat;
    mat.SetDims(m, m);
    Mat<zz_p> array[s];
#if defined(DEBUG)
    cout << "main step 0.2" << endl;
#endif
    vector<vector<int> >to;
    to.resize(max_tvalue + 1);
#if defined(DEBUG)
    cout << "main step 1" << endl;
#endif
    for (int i = 0; i <= max_tvalue; i++) {
        to[i].resize(max);
        for (int64_t j = 0; j < max; j++) {
            to[i][j] = 0;
        }
    }
#if defined(DEBUG)
    cout << "main step 2" << endl;
#endif
    for (int64_t cnt = 1; cnt < max; cnt++) {
        if (!make_mat(mat, cnt)) {
            continue;
        }
        int order = calc_order(mat, max - 1);
        if (order < s) {
            continue;
        }
        array[0] = mat;
        for (int j = 1; j < s; j++) {
            array[j] = array[j - 1] * mat;
        }
        int tvalue = calc_tvalue(array, s, max_tvalue);
        if (tvalue > max_tvalue || tvalue < 0) {
            continue;
        }
        to[tvalue][order] += 1;
        if (tvalue == 0 && order == max -1) {
            cout << "found!" << endl;
            cout << "cnt = " << dec << cnt << endl;
            cout << "tvalue = " << dec << tvalue << endl;
            cout << "order = " << dec << order << endl;
            cout << mat;
        }
    }
    cout << "(tvalue, order) = cnt" << endl;
    for (int i = 0; i <= max_tvalue; i++) {
        for (int64_t j = 0; j < max; j++) {
            if (to[i][j] == 0) {
                continue;
            }
            cout << "(" << dec << i << ","
                 << j << ") = " << to[i][j] << endl;
        }
    }
    return 0;
}

namespace {
    bool make_mat(Mat<zz_p>& mat, uint64_t num)
    {
#if defined(DEBUG)
        cout << "start make_mat" << endl;
#endif
        //cout << "num = " << dec << num << endl;
        int p = zz_p::modulus();
        int m = mat.NumRows();
        //const uint64_t mask = ~UINT64_C(0) >> (64 - m);
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < m; j++) {
                zz_p z(num % p);
                num = num / p;
                mat.put(i, j, z);
            }
        }
        //cout << "make_mat 1:" << mat << endl;
        Mat<zz_p> w = mat;
        if (gauss(w) != m) {
#if defined(DEBUG)
            cout << "end make_mat" << endl;
#endif
            return false;
        }
        //cout << "make_mat 2:\n" << mat << endl;
#if defined(DEBUG)
        cout << "end make_mat" << endl;
#endif
        return true;
    }

    size_t calc_order(Mat<zz_p>& mat, size_t max)
    {
        Mat<zz_p> w = mat;
        for (size_t i = 1; i <= max; i++) {
            w = w * mat;
            if (w == mat) {
                return i;
            }
        }
        return max;
    }
}


src/calc_tvalue_fp.hpp
#pragma once
#ifndef CALC_TVALUE_FP_HPP
#define CALC_TVALUE_FP_HPP

//#include <NTL/matrix.h>
#include <NTL/lzz_p.h>
#include <NTL/mat_lzz_p.h>
#include "combination.hpp"

int calc_tvalue(NTL::Mat<NTL::zz_p> array[], int dim, int max_tvalue)
{
    using namespace std;
    using namespace NTL;

    uint32_t s = dim;
    uint32_t m = array[0].NumRows();
    Combination index(s);
    int t = 0;
    for (; t <= max_tvalue; t++) {
        index.reset(m - t);
#if defined(DEBUG)
        cout << "t = " << dec << t << endl;
#endif
        //vector<uint32_t> v;
        bool pass = true;
        for (;;) {
            Mat<zz_p> v;
            v.SetDims(m - t, m);
#if defined(DEBUG)
            index.print(cout);
#endif
            int vrow = 0;
            for (uint32_t i = 0; i < s; ++i) {
                for (int j = 0; j < index[i]; ++j) {
                    v[vrow++] = array[i][j];
                }
            }
            if (gauss(v) != m-t) {
                pass = false;
                break;
            }
            bool hasNext = index.next();
            if (!hasNext) {
                break;
            }
        }
        if (pass) {
            return t;
        }
    }
    return -1;
}

#endif // CALC_TVALUE_FP_HPP


src/order_test_I.cpp
#include <inttypes.h>
#include <iostream>
#include <iomanip>
#include <cstdlib>
#include <cerrno>
#include <NTL/mat_GF2.h>
#include <NTL/GF2.h>
#include <NTL/matrix.h>
#include <vector>
#include "calc_tvalue.hpp"

using namespace std;
using namespace NTL;

//#define DEBUG

namespace {
    Mat<GF2> Ident;
    bool make_mat(Mat<GF2>& mat, uint64_t num);
    size_t calc_order(Mat<GF2>& mat, size_t max);
}

int main(int argc, char * argv[])
{

    if (argc < 4) {
        cout << argv[0] << " s m max_tvalue" << endl;
        return -1;
    }
    errno = 0;
    int s = strtoul(argv[1], NULL, 10);
    int m = strtoul(argv[2], NULL, 10);
    int max_tvalue = strtoul(argv[3], NULL, 10);
    if (errno) {
        cout << " s, m and max_tvalue should be numbers" << endl;
        return -1;
    }
    cout << "# order_test_I I B B^2 ... B^s-1" << endl;
    cout << "# s = " << dec << s << endl;
    cout << "# m = " << dec << m << endl;
    cout << "# max_tvalue = " << dec << max_tvalue << endl;
    ident(Ident, m);
#if defined(DEBUG)
    cout << "main step 0.1" << endl;
#endif
    int64_t max = 1 << (m * m);
    Mat<GF2> mat;
    mat.SetDims(m, m);
    Mat<GF2> array[s];
#if defined(DEBUG)
    cout << "main step 0.2" << endl;
#endif
    vector<vector<int> >to;
    to.resize(max_tvalue + 1);
#if defined(DEBUG)
    cout << "main step 1" << endl;
#endif
    for (int i = 0; i <= max_tvalue; i++) {
        to[i].resize(max);
        for (int64_t j = 0; j < max; j++) {
            to[i][j] = 0;
        }
    }
#if defined(DEBUG)
    cout << "main step 2" << endl;
#endif
    for (int64_t cnt = 1; cnt < max; cnt++) {
        if (!make_mat(mat, cnt)) {
            continue;
        }
        int order = calc_order(mat, max - 1);
        if (order < s) {
            continue;
        }
        array[0] = Ident;
        for (int j = 1; j < s; j++) {
            array[j] = array[j - 1] * mat;
        }
        int tvalue = calc_tvalue(array, s, max_tvalue);
        if (tvalue > max_tvalue || tvalue < 0) {
            continue;
        }
        to[tvalue][order] += 1;
        if (tvalue == 0 && order == max -1) {
            cout << "found!" << endl;
            cout << "cnt = " << dec << cnt << endl;
            cout << "tvalue = " << dec << tvalue << endl;
            cout << "order = " << dec << order << endl;
            cout << mat;
        }
    }
    cout << "(tvalue, order) = cnt" << endl;
    for (int i = 0; i <= max_tvalue; i++) {
        for (int64_t j = 0; j < max; j++) {
            if (to[i][j] == 0) {
                continue;
            }
            cout << "(" << dec << i << ","
                 << j << ") = " << to[i][j] << endl;
        }
    }
    return 0;
}

namespace {
    bool make_mat(Mat<GF2>& mat, uint64_t num)
    {
#if defined(DEBUG)
        cout << "start make_mat" << endl;
#endif
        //cout << "num = " << dec << num << endl;
        int m = mat.NumRows();
        const uint64_t mask = ~UINT64_C(0) >> (64 - m);
        for (int i = 0; i < m; i++) {
            uint64_t row = num & mask;
            if (row == 0) {
#if defined(DEBUG)
                cout << "end make_mat" << endl;
#endif
                return false;
            }
            for (int j = 0; j < m; j++) {
                //cout << "mat.put(" << i << "," << j << "," << (row & 1) << ")"
                //     << endl;
                mat.put(i, j, row & 1);
                row = row >> 1;
            }
            num = num >> m;
        }
        //cout << "make_mat 1:" << mat << endl;
        Mat<GF2> w = mat;
        if (gauss(w) != m) {
#if defined(DEBUG)
            cout << "end make_mat" << endl;
#endif
            return false;
        }
        //cout << "make_mat 2:\n" << mat << endl;
#if defined(DEBUG)
        cout << "end make_mat" << endl;
#endif
        return true;
    }

    size_t calc_order(Mat<GF2>& mat, size_t max)
    {
        Mat<GF2> w = mat;
        for (size_t i = 1; i <= max; i++) {
            w = w * mat;
            if (w == mat) {
                return i;
            }
        }
        return max;
    }
}
